src/core/main.c

```diff
@@ -1666,8 +1666,27 @@ int main(int argc, char *argv[]) {
                 test_usr();
         }
 
-        if (arg_system && arg_runtime_watchdog > 0 && arg_runtime_watchdog != USEC_INFINITY)
+        if (arg_system && arg_runtime_watchdog > 0 && arg_runtime_watchdog != USEC_INFINITY) {
                 watchdog_set_timeout(&arg_runtime_watchdog);
+        } else if(!arg_system) {
+                // we are a user-slice triggering system-slice's service-watchdog.
+                char * new_time_str = getenv("WATCHDOG_USEC");
+                if(new_time_str) {
+                        int new_time;
+                        char * p;
+                        new_time = strtoul(new_time_str, &p, 10);
+                        if(0 == *p) {
+                                watchdog_enable_user_mode();
+                                arg_runtime_watchdog = new_time;
+                                log_debug(PACKAGE_STRING " setting systemd user slice watchdog time to %d us", arg_runtime_watchdog);
+                                watchdog_set_timeout(&arg_runtime_watchdog);
+                        } else {
+                                arg_runtime_watchdog = 0;
+                        }
+                } else {
+                        arg_runtime_watchdog = 0;
+                }
+        }
 
```

Comment:
Please have a look at CODING_STYLE, there are many issues with the coding style used here, please make sure to submit code that is in line with CODING_STYLE.

That said, I am pretty sure we should simply use sd-event's watchdog support for this. i.e. sd-event already can send out WATCHDOG=1 notification messages on its own, see sd_event_set_watchdog(3) man page, and we really should use this here, too.